store/tikv/mocktikv/aggregate.go

```diff
@@ -0,0 +1,302 @@
+	if e.aggCtxsMap == nil {
+		e.aggCtxsMap = make(aggCtxsMapper, 0)
+	}
+	if !e.executed {
+		for {
+			hasMore, err := e.innerNext(goCtx)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if !hasMore {
+				break
+			}
+		}
+		e.executed = true
+	}
+
+	if e.currGroupIdx >= len(e.groups) {
+		return nil, nil
+	}
+	gk := e.groupKeys[e.currGroupIdx]
+	value = make([][]byte, 0, len(e.groupByExprs)+2*len(e.aggExprs))
+	aggCtxs := e.getContexts(gk)
+	for i, agg := range e.aggExprs {
+		partialResults := agg.GetPartialResult(aggCtxs[i])
+		for _, result := range partialResults {
+			data, err := codec.EncodeValue(nil, result)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			value = append(value, data)
+		}
+	}
+	value = append(value, e.groupKeyRows[e.currGroupIdx]...)
+	e.currGroupIdx++
+
+	return value, nil
+}
+
+func (e *hashAggExec) getGroupKey() ([]byte, [][]byte, error) {
+	length := len(e.groupByExprs)
+	if length == 0 {
+		return nil, nil, nil
+	}
+	bufLen := 0
+	row := make([][]byte, 0, length)
+	for _, item := range e.groupByExprs {
+		v, err := item.Eval(e.row)
+		if err != nil {
+			return nil, nil, errors.Trace(err)
+		}
+		b, err := codec.EncodeValue(nil, v)
+		if err != nil {
+			return nil, nil, errors.Trace(err)
+		}
+		bufLen += len(b)
+		row = append(row, b)
+	}
+	buf := make([]byte, 0, bufLen)
+	for _, col := range row {
+		buf = append(buf, col...)
+	}
+	return buf, row, nil
+}
+
+// aggregate updates aggregate functions with row.
+func (e *hashAggExec) aggregate(value [][]byte) error {
+	err := e.evalCtx.decodeRelatedColumnVals(e.relatedColOffsets, value, e.row)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	// Get group key.
+	gk, gbyKeyRow, err := e.getGroupKey()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if _, ok := e.groups[string(gk)]; !ok {
+		e.groups[string(gk)] = struct{}{}
+		e.groupKeys = append(e.groupKeys, gk)
+		e.groupKeyRows = append(e.groupKeyRows, gbyKeyRow)
+	}
+	// Update aggregate expressions.
+	aggCtxs := e.getContexts(gk)
+	for i, agg := range e.aggExprs {
+		err = agg.Update(aggCtxs[i], e.evalCtx.sc, e.row)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (e *hashAggExec) getContexts(groupKey []byte) []*aggregation.AggEvaluateContext {
+	groupKeyString := string(groupKey)
+	aggCtxs, ok := e.aggCtxsMap[groupKeyString]
+	if !ok {
+		aggCtxs = make([]*aggregation.AggEvaluateContext, 0, len(e.aggExprs))
+		for _, agg := range e.aggExprs {
+			aggCtxs = append(aggCtxs, agg.CreateContext())
+		}
+		e.aggCtxsMap[groupKeyString] = aggCtxs
+	}
+	return aggCtxs
+}
+
+type streamAggExec struct {
+	evalCtx           *evalContext
+	aggExprs          []aggregation.Aggregation
+	aggCtxs           []*aggregation.AggEvaluateContext
+	groupByExprs      []expression.Expression
+	relatedColOffsets []int
+	row               types.DatumRow
+	tmpGroupByRow     types.DatumRow
+	currGroupByRow    types.DatumRow
```

Comment:
I think currGroupByRow should be previous gby row, and next Group by row should be current gby row,